### PR TITLE
mongodb: init at 4.4.10, init at 5.0.3, refactor and update tests

### PR DESCRIPTION
**Pull request by bryanasdev000**
###### Motivation for this change

- init release 4.4 at 4.4.10
- init release 5.0 at 5.0.3
- refactor for new releases introduced
- update tests to make use of the new package versions

This is a WIP since I can't make it build in this current state (it can affect the old releases too), also the workflow is a bit slow here since building mongo takes around an hour.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/nosql/mongodb/v5_0.nix [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{ stdenv, callPackage, lib, sasl, boost, Security, CoreFoundation, cctools }:
+
+let
+  buildMongoDB = callPackage ./mongodb.nix {
+    inherit sasl;
+    inherit boost;
+    inherit Security;
+    inherit CoreFoundation;
+    inherit cctools;
```
Comment (SuperSandro2000) [inline pkgs/servers/nosql/mongodb/v5_0.nix:9]:
```suggestion
    inherit sasl boost Security CoreFoundation cctools;
```
this could be simplified to this.

**Comment by andersk** [comment]
This seems to have been superseded by

- #137939

although that also doesn’t work:

```console
$ NIXPKGS_ALLOW_UNFREE=1 nix log --impure nixpkgs#mongodb-4_4
[…]
In file included from src/mongo/db/pipeline/pipeline.h:39,
                 from src/mongo/db/pipeline/sharded_agg_helpers.h:32,
                 from src/mongo/db/pipeline/sharded_agg_helpers.cpp:33:
src/mongo/db/exec/plan_stats.h:214:10: error: ‘optional’ in namespace ‘std’ does not name a template type
  214 |     std::optional<std::string> replanReason;
      |          ^~~~~~~~
src/mongo/db/exec/plan_stats.h:41:1: note: ‘std::optional’ is defined in header ‘<optional>’; did you forget to ‘#include <optional>’?
   40 | #include "mongo/util/container_size_helper.h"
  +++ |+#include <optional>
   41 | #include "mongo/util/time_support.h"
[…]
scons: building terminated because of errors.
build/nixos/mongo/db/pipeline/sharded_agg_helpers.o failed: Error 1

$ NIXPKGS_ALLOW_UNFREE=1 nix log --impure nixpkgs#mongodb-5_0
[…]
Checking for SSLeay_version(0) in C library crypto... no
Couldn't find OpenSSL crypto.h header and library
See /build/mongodb-src-r5.0.7/build/scons/config.log for details
```

**Comment by bryanasdev000** [comment]
> This seems to have been superseded by
> 
> * [Update mongodb to 4.4/5.0 #137939](https://github.com/NixOS/nixpkgs/pull/137939)
> 
> although that also doesn’t work:
> 
> ```
> $ NIXPKGS_ALLOW_UNFREE=1 nix log --impure nixpkgs#mongodb-4_4
> […]
> In file included from src/mongo/db/pipeline/pipeline.h:39,
>                  from src/mongo/db/pipeline/sharded_agg_helpers.h:32,
>                  from src/mongo/db/pipeline/sharded_agg_helpers.cpp:33:
> src/mongo/db/exec/plan_stats.h:214:10: error: ‘optional’ in namespace ‘std’ does not name a template type
>   214 |     std::optional<std::string> replanReason;
>       |          ^~~~~~~~
> src/mongo/db/exec/plan_stats.h:41:1: note: ‘std::optional’ is defined in header ‘<optional>’; did you forget to ‘#include <optional>’?
>    40 | #include "mongo/util/container_size_helper.h"
>   +++ |+#include <optional>
>    41 | #include "mongo/util/time_support.h"
> […]
> scons: building terminated because of errors.
> build/nixos/mongo/db/pipeline/sharded_agg_helpers.o failed: Error 1
> 
> $ NIXPKGS_ALLOW_UNFREE=1 nix log --impure nixpkgs#mongodb-5_0
> […]
> Checking for SSLeay_version(0) in C library crypto... no
> Couldn't find OpenSSL crypto.h header and library
> See /build/mongodb-src-r5.0.7/build/scons/config.log for details
> ```

Thanks for the tip, been busy and I did not give enough attention to this PR :P 

Gonna test!

**Comment by bryanasdev000** [comment]
> This seems to have been superseded by
> 
> * [Update mongodb to 4.4/5.0 #137939](https://github.com/NixOS/nixpkgs/pull/137939)
> 
> although that also doesn’t work:
> 
> ```
> $ NIXPKGS_ALLOW_UNFREE=1 nix log --impure nixpkgs#mongodb-4_4
> […]
> In file included from src/mongo/db/pipeline/pipeline.h:39,
>                  from src/mongo/db/pipeline/sharded_agg_helpers.h:32,
>                  from src/mongo/db/pipeline/sharded_agg_helpers.cpp:33:
> src/mongo/db/exec/plan_stats.h:214:10: error: ‘optional’ in namespace ‘std’ does not name a template type
>   214 |     std::optional<std::string> replanReason;
>       |          ^~~~~~~~
> src/mongo/db/exec/plan_stats.h:41:1: note: ‘std::optional’ is defined in header ‘<optional>’; did you forget to ‘#include <optional>’?
>    40 | #include "mongo/util/container_size_helper.h"
>   +++ |+#include <optional>
>    41 | #include "mongo/util/time_support.h"
> […]
> scons: building terminated because of errors.
> build/nixos/mongo/db/pipeline/sharded_agg_helpers.o failed: Error 1
> 
> $ NIXPKGS_ALLOW_UNFREE=1 nix log --impure nixpkgs#mongodb-5_0
> […]
> Checking for SSLeay_version(0) in C library crypto... no
> Couldn't find OpenSSL crypto.h header and library
> See /build/mongodb-src-r5.0.7/build/scons/config.log for details
> ```

Indeed, similar results here...

**Comment by bryanasdev000** [comment]
I did a git checkout to the original commit of the PR and was able to start compilation of MongoDB v5, did not finish because well my machine was burning itself. Same for v4.4.

Not sure what happens in v4, but for v5 my guess is an openssl update.

Going to open an issue for follow up.

**Comment by bryanasdev000** [comment]
Here https://github.com/NixOS/nixpkgs/issues/171928

**Pull request closed without merging**